File: scripts/enemies/basic_flying.py
import random

from scripts.enemies._enemy import Enemy


class BasicFlying(Enemy):
    size_range = (20, 50)
    costume = "1"
    health_factor = 0.5

    def __init__(self, default_zooms, color, size, nest_health, x, y):
        super().__init__(default_zooms, BasicFlying.costume, color, x, y, size, nest_health * BasicFlying.health_factor)
        self.knockback = 0.1
        self.speed = 1.5
        self.knockback_resistance = 0.8
        self.gravity_multiplier = 0

    def tick_enemy_behavior(self, frame_length, player):
        if self.mode == "Walk":
            if abs(player.x - self.x) > self.size / 2 or abs(player.y - self.y) > self.size / 2:
                rand = random.randint(0, 3)
                if (player.x < self.x and rand != 3) or rand == 0:
                    self.x_speed -= 0.0003 * frame_length * self.speed
                else:
                    self.x_speed += 0.0003 * frame_length * self.speed
                rand = random.randint(0, 3)
                if (player.y < self.y and rand != 3) or rand == 0:
                    self.y_speed -= 0.0003 * frame_length * self.speed
                else:
                    self.y_speed += 0.0003 * frame_length * self.speed
                self.x_speed *= 0.995**frame_length
                self.y_speed *= 0.995**frame_length
            else:
                self.mode = "Attack"
                self.animation_timer = 0

File: cpp_port/BasicFlying.cpp
#include "BasicFlying.h"
#include "Util.h"
#include <cmath>

BasicFlying::BasicFlying(SDL_Renderer* renderer, Color color, double size, double nestHealth, double x, double y)
    : Enemy(renderer, kCostume, color, x, y, size, nestHealth * kHealthFactor) {
    knockback = 0.1;
    speed = 1.5;
    knockbackResistance = 0.8;
    gravityMultiplier = 0.0;
}

// def tick_enemy_behavior(self, frame_length, player): ...
void BasicFlying::tickEnemyBehavior(double frameLength, const EnemyPlayerView& player) {
    if (mode != EnemyMode::Walk) return;

    if (std::abs(player.x - x) > size / 2.0 || std::abs(player.y - y) > size / 2.0) {
        int rndX = Util::randint(0, 3);
        if ((player.x < x && rndX != 3) || rndX == 0) xSpeed -= 0.0003 * frameLength * speed;
        else xSpeed += 0.0003 * frameLength * speed;

        int rndY = Util::randint(0, 3);
        if ((player.y < y && rndY != 3) || rndY == 0) ySpeed -= 0.0003 * frameLength * speed;
        else ySpeed += 0.0003 * frameLength * speed;

        xSpeed *= std::pow(0.995, frameLength);
        ySpeed *= std::pow(0.995, frameLength);
    } else {
        mode = EnemyMode::Attack;
        animationTimer_ = 0;
    }
}
